### PR TITLE
Added Reversible Instance Normalization

### LINKED ISSUE #1121: Reversible Instance Normalization for Accurate Time-Series Forecasting against Distribution Shift.
I've seen pretty good results using [RevIN](https://github.com/ts-kim/RevIN) with larger datasets. I'm also waiting on a license.

### LINKED ISSUE #1861: Add Reversible Instance Normalization to Torch Models
Link to paper: https://openreview.net/forum?id=cGDAkQo1C0p

I have been implementing TSMixer (#1807) and noted that it was using RevIN (Reversible Instance Normalization).

I went to the original source paper to understand the purpose of it and it seems to obtain fantastic results as essentially a preprocessing step within the models themselves that helps address shifting distributions. The paper shows improved performance on Informer, N-BEATS, and SCINet (#1860).

image

Using my implementation and plugging it into N-HiTS, I was able to improve performance by a good marging using the Energy dataset.

with:
######################################
MSE:  0.2514766273597132
MAE:  0.334557913679719

without:
######################################
MSE:  0.30861193252719554
MAE:  0.41421373603271244

My main questions are around implementation. Can we add a wrapper around the forward method in the Torch-based models? Is there some other way to add this feature without having to edit a ton of the torch models?

### PR DESCRIPTION
**Pull request by alexcolpitts96**
Fixes #1121, fixes #1861.

Adds RIN as a normalization layer option requested in #1121. 

This is a prerequisite component for #1861 and #1807.

### CODE REVIEW

File: darts/models/components/layer_norm_variants.py [hunk 1]
```diff
@@ -54,3 +54,82 @@ def __init__(self, input_size, **kwargs):
...
+
+            massage_shape = not (
+                (self.axis is not -1) and (self.axis is not x.ndim - 1)
+            )
+
+            # if axis isn't the last dimension, swap it to the last dimension
+            if massage_shape:
+                x = x.swapaxes(-2, self.axis)
+
+            x = x - self.affine_bias[target_slice]
```
Comment (dennisbader) [inline darts/models/components/layer_norm_variants.py:126]: could you add a comment why target_slice is needed?

File: darts/models/components/layer_norm_variants.py [hunk 2]
```diff
@@ -54,3 +54,82 @@ def __init__(self, input_size, **kwargs):
...
+            x = x / self.affine_weight[target_slice]
```
Comment (dennisbader) [inline darts/models/components/layer_norm_variants.py:127]: official implementation uses:
```
x = x / (self.affine_weight[target_slice] + self.eps*self.eps)
```
did you intentionally drop the `eps` part?

File: darts/tests/models/components/test_layer_norm_variants.py [hunk 3]
```diff
@@ -25,3 +28,37 @@ def test_lnv(self):
...
+                affine, axis, input_dim = run
+
+                if (x.shape[axis] is input_dim) and affine:
+
+                    with self.assertRaises(RuntimeError):
+                        rin = ReversibleInstanceNorm(
+                            axis=axis, input_dim=input_dim, affine=affine
+                        )
+                        x_norm = rin(x, "norm")
+                        x_denorm = rin(x_norm, "denorm")
```
Comment (dennisbader) [inline darts/tests/models/components/test_layer_norm_variants.py:55]: it's better to only have the line which should raise the Error inside the with body.
The Norm creation should not raise an error. Will the error always be raised in the `x_norm` line?

Then we could rewrite as:
```suggestion
                    rin = ReversibleInstanceNorm(
                        axis=axis, input_dim=input_dim, affine=affine
                    )
                    with self.assertRaises(RuntimeError):
                        _ = rin(x, "norm")
                    with self.assertRaises(RuntimeError):
                        _ = rin(x, "denorm")
```

File: darts/models/components/layer_norm_variants.py [hunk 4]
```diff
@@ -54,3 +54,82 @@ def __init__(self, input_size, **kwargs):
 class LayerNorm(nn.LayerNorm):
     def __init__(self, input_size, **kwargs) -> None:
         super().__init__(input_size, **kwargs)
+
+
+class ReversibleInstanceNorm(nn.Module):
```
Comment (dennisbader) [inline darts/models/components/layer_norm_variants.py:59]: I wonder if `RINorm` or `RevINorm` would be better names, as for longer ones such as RMSNorm we also use the abbreviation.
what do you think?

File: darts/models/components/layer_norm_variants.py
Comment (dennisbader) [inline darts/models/components/layer_norm_variants.py:1]: How should/can the `ReversibleInstanceNorm` be used in practice (in Darts)?

The paper shows very nice results also for NBEATS. Do you have something in mind how we could integrate it to these models?

Also, can it be used in all the places as the other layer norms here? For example as `norm_type` in `TFTModel` and `TransformerModel`? If so, then we should add it to the documentation of these models.

File: darts/tests/models/components/test_layer_norm_variants.py [hunk 5]
```diff
@@ -25,3 +28,37 @@ def test_lnv(self):
                 ln = layer_norm(4)
                 inputs = torch.zeros(1, 4, 4)
                 ln(inputs)
+
+        def test_rin(self):
+
+            x = torch.randn(3, 4, 7)
```
Comment (dennisbader) [inline darts/tests/models/components/test_layer_norm_variants.py:34]: this test is subjected to randomness and can eventually fail. Can we address this?

**Review by dennisbader — CHANGES_REQUESTED** [review]
Very nice, thanks a lot for this very cool addition @alexcolpitts96 🚀 :)

I only had some minor suggestions and general questions about small deviations from the official implementation and the usage of the Norm in Darts

File: darts/models/components/layer_norm_variants.py
Comment (alexcolpitts96) [reply darts/models/components/layer_norm_variants.py:1]: I am not sure how to apply it for probabilistic models, but I was thinking there might be some way to wrap the forward function of a model. When I added it to NHiTS it was a simple as adding RIN to __init__ and calling transform before/after. Maybe add an argument to the torch model base class, "use_reversible_instance_norm"?

For covariates and probabilistic forecasts there will need to be some adjustments, but I expect it to be possible. Ignoring covariates on the denorm is easy, but denorming more samples is not something I can immediately think of a robust solution for.

File: darts/models/components/layer_norm_variants.py (hunk 4 above)
Comment (alexcolpitts96) [reply darts/models/components/layer_norm_variants.py:59]: I had been writing RIN, but RINorm is cleaner and feels more in line with RMSNorm.

File: darts/models/components/layer_norm_variants.py [hunk 6]
```diff
@@ -54,3 +54,82 @@ def __init__(self, input_size, **kwargs):
...
+            x = self._normalize(x)
+        elif mode == "denorm":
+            x = self._denormalize(x, target_slice)
+        else:
+            raise NotImplementedError
+        return x
+
+    def _get_statistics(self, x):
+        self.mean = torch.mean(x, dim=self.axis, keepdim=True)
+        self.std = torch.sqrt(torch.var(x, dim=self.axis, keepdim=True) + self.eps)
```
[the remark this answers is not in the document]
Comment (alexcolpitts96) [reply darts/models/components/layer_norm_variants.py:91]: Totally missed those. Added them back in.

File: darts/tests/models/components/test_layer_norm_variants.py (hunk 5 above)
Comment (alexcolpitts96) [reply darts/tests/models/components/test_layer_norm_variants.py:34]: I added RNG seeding for torch and numpy.

File: darts/models/components/layer_norm_variants.py (hunk 2 above)
Comment (alexcolpitts96) [reply darts/models/components/layer_norm_variants.py:127]: I have no idea how I missed that. I added it.

File: darts/models/components/layer_norm_variants.py [hunk 7]
```diff
@@ -54,3 +54,82 @@ def __init__(self, input_size, **kwargs):
...
+        x = x / self.std
+        if self.affine:
+
+            massage_shape = not (
+                (self.axis is not -1) and (self.axis is not x.ndim - 1)
+            )
+
+            # if axis isn't the last dimension, swap it to the last dimension
+            if massage_shape:
+                x = x.swapaxes(-2, self.axis)
```
[the remark this answers is not in the document]
Comment (alexcolpitts96) [reply darts/models/components/layer_norm_variants.py:120]: I seem to recall there was an issue with the matrix multiplication not playing nice if the normalization axis wasn't the last dimension. 

I added tests to cover this because it was a problem. The original source code I was basing my implementation off of assumed the time and feature dimensions were swapped. I will double check this, but it should be correct, but there may be a faster implementation.

File: CHANGELOG.md [hunk 8]
```diff
@@ -19,6 +19,7 @@ but cannot always guarantee backwards compatibility. Changes that may **break co
   - Model creation parameter `forecasting_models` now supports a mix of `LocalForecastingModel` and `GlobalForecastingModel` (single `TimeSeries` training/inference only, due to the local models). [#1745](https://github.com/unit8co/darts/pull/1745) by [Antoine Madrona](https://github.com/madtoinou).
   - Future and past covariates can now be used even if `forecasting_models` have different covariates support. The covariates passed to `fit()`/`predict()` are used only by models that support it. [#1745](https://github.com/unit8co/darts/pull/1745) by [Antoine Madrona](https://github.com/madtoinou).
   - `RegressionEnsembleModel` and `NaiveEnsembleModel` can generate probabilistic forecasts, probabilistics `forecasting_models` can be sampled to train the `regression_model`, updated the documentation (stacking technique). [#1692](https://github.com/unit8co/darts/pull/#1692) by [Antoine Madrona](https://github.com/madtoinou).
+  - Added `RINorm` as a new layer normalization option. [#1121](https://github.com/unit8co/darts/issues/1121) by [Alex Colpitts](https://github.com/alexcolpitts96)
```
Comment (dennisbader) [inline CHANGELOG.md:22]:
```suggestion
  - Added `RINorm` (Reversible Instance Norm) as a new layer normalization option. [#1121](https://github.com/unit8co/darts/issues/1121) by [Alex Colpitts](https://github.com/alexcolpitts96)
```

File: darts/tests/models/components/test_layer_norm_variants.py [hunk 9]
```diff
@@ -25,3 +28,24 @@ def test_lnv(self):
...
+            torch.manual_seed(42)
+
+            x = torch.randn(3, 4, 7)
+            affine_options = [True, False]
+
+            # test with and without affine and correct input dim
+            for affine in affine_options:
+
+                rin = RINorm(input_dim=7, affine=affine)
+                x_norm = rin.forward(x)
```
Comment (dennisbader) [inline darts/tests/models/components/test_layer_norm_variants.py:44]:
```suggestion
                x_norm = rin(x)
```

**Review by dennisbader — CHANGES_REQUESTED** [review]
Nice, thanks @alexcolpitts96 🚀 
Only had some last minor suggestions.
After that we can merge :)

File: darts/models/components/layer_norm_variants.py [hunk 10]
```diff
@@ -54,3 +54,55 @@ def __init__(self, input_size, **kwargs):
...
+        self.input_dim = input_dim
+        self.eps = eps
+        self.affine = affine
+
+        if self.affine:
+            self.affine_weight = nn.Parameter(torch.ones(self.input_dim))
+            self.affine_bias = nn.Parameter(torch.zeros(self.input_dim))
+
+    def forward(self, x):
+        calc_dims = tuple(range(1, x.ndim - 1))
```
[the remark this answers is not in the document]
Comment (alexcolpitts96) [reply darts/models/components/layer_norm_variants.py:91]: I added some comments. I want to keep it as taking a torch.Tensor for now and not a specific data set. I may change this when it comes to integrating it with models since it may make covariate handling very gross in the models. 

When I add it to TiDE (and maybe NHiTS for testing) I will see how it looks and update it accordingly with another PR.

**Review by dennisbader — APPROVED** [review]
Thanks a lot @alexcolpitts96, everything is looking great now! 💯 
Will merge as soon as all tests passed.

**Comment by gdevos010** [comment]
@dennisbader @alexcolpitts96 I haven't had a chance to do a deep dive into the code changes or what TIDE/TSMixer model requires, but since this MR closed my issue, I thought I would chime in on my understanding of RevIN. 

RevIN is not a layer normalization.  You can see in the [Informer](https://github.com/ts-kim/RevIN/blob/34147dfabda0e0b9a4f3e78e30bd56ed1fccbf82/baselines/Informer2020/models/model.py#L77-L96) example and it implemented in [PatchTST](https://github.com/PatchTST/PatchTST/blob/bb0bc6058ddc421c02e8afe77e7e8db99f913957/PatchTST_supervised/layers/PatchTST_backbone.py#L60-L82) that it is applied at the start and end of the `forward()` method, not in-between the layers as layer norms are. To reduce the confusion, I would move it to a separate file since having it in `layer_norm_variants.py` does matter when doing we do the layer norm lookups in some of the models. 

The easiest way to implement REVIN is to add it `PLForecastingModule` as recommended [here ](https://github.com/unit8co/darts/issues/1861#issuecomment-1611137155) so that all models have access to RevIN. To me, #1121 is not completed until its added to `PLForecastingModule`. I see RevIN included in many of the new forecasting models, so I appreciate the effort to include it.

**Comment by dennisbader** [comment]
You're right @gdevos010. We should add it to another file than the layer norm variants. So far RevIN is not used by any model, but @alexcolpitts96 mentioned integrating it while working on the other PRs.
I'll reopen your issue so we won't forget.

For non-probabilistic models (not using a likelihood) it will work fine at the beginning and end of the forward pass. We still have to check how to best do it for models using a likelihood.
